File: packages/core/src/ranking.test.ts
import { describe, expect, test } from "bun:test";
import {
  DEFAULT_PARTICIPATION,
  deployableUsd,
  expectedWeeklyUsd,
  type PairEvidence,
  rankByExpectedDollars,
  score,
  shouldExit,
  shouldSwitch,
  switchBand,
} from "./ranking";

/** The measured shape of a well-evidenced pair, so each test varies only what it is about. */
const evidence = (over: Partial<PairEvidence>): PairEvidence => ({
  ratePerWeek: 0.003, // ~15.6% APR, the measured size-weighted figure
  chargeDays: 30,
  thinnerLegOiUsd: 25_000_000,
  ...over,
});

/** The pool prior used throughout: the measured median of $3.12 per $10k per week. */
const PRIOR = 0.000312;

describe("score", () => {
  test("shrinks a thin sample toward the prior, and leaves a thick one nearly alone", () => {
    // Same observation, different evidence behind it. The sparse pair keeps a usable figure rather
    // than being discarded, and still loses to the well-evidenced one -- the point of k.
    const sparse = score(evidence({ ratePerWeek: 0.01, chargeDays: 4 }), PRIOR);
    const thick = score(evidence({ ratePerWeek: 0.01, chargeDays: 30 }), PRIOR);

    expect(sparse).toBeCloseTo((4 * 0.01 + 10 * PRIOR) / 14, 8);
    expect(thick).toBeCloseTo((30 * 0.01 + 10 * PRIOR) / 40, 8);
    expect(sparse).toBeLessThan(thick);
    // And the sparse one is pulled most of the way back toward the prior.
    expect(sparse).toBeLessThan(0.01 * 0.75);
  });

  test("no evidence at all scores exactly the prior", () => {
    expect(score(evidence({ chargeDays: 0 }), PRIOR)).toBeCloseTo(PRIOR, 10);
  });

  test("the prior is the pool median, not zero", () => {
    // Zero would claim an unobserved pair pays nothing, which is not what ignorance means. A pair
    // with one charging day sits near the median rather than near zero.
    expect(score(evidence({ ratePerWeek: 0, chargeDays: 1 }), PRIOR)).toBeGreaterThan(0);
  });

  test("NaN in, NaN out -- never a silent zero that would rank as merely mediocre", () => {
    expect(score(evidence({ ratePerWeek: Number.NaN }), PRIOR)).toBeNaN();
  });
});

describe("switchBand", () => {
  test("a switch must at least pay for itself over the holding period", () => {
    // 20 bps round trip (four fills at 5 bps), held four weeks.
    expect(switchBand(0.002, 4)).toBeCloseTo(0.0005, 10);
  });

  test("the band is a meaningful fraction of the weekly rate, not a rounding error", () => {
    // Against the measured 0.003/week, a 0.0005 band is about a sixth of the rate. That is real
    // hysteresis: this is what stops the 71% overnight churn.
    const band = switchBand(0.002, 4);
    expect(band / 0.003).toBeGreaterThan(0.1);
    expect(band / 0.003).toBeLessThan(0.3);
  });

  test("a better fee tier lowers the bar to switching", () => {
    // A VIP book pays less to move, so it can afford to chase smaller improvements.
    expect(switchBand(0.0006, 4)).toBeLessThan(switchBand(0.002, 4));
  });

  test("the cube-root term is off until calibrated", () => {
    // Shipping an uncalibrated constant would be exactly the fitting the design forbids, so the
    // default reduces to pure payback, which is arithmetic rather than a fitted parameter.
    expect(switchBand(0.002, 4)).toBe(switchBand(0.002, 4, 0));
    expect(switchBand(0.002, 4, 0.01)).toBeGreaterThan(switchBand(0.002, 4, 0));
  });

  test("no cost means no band", () => {
    expect(switchBand(0)).toBe(0);
    expect(switchBand(Number.NaN)).toBe(0);
  });
});

describe("shouldSwitch", () => {
  const band = switchBand(0.002, 4); // 0.0005

  test("holds when the improvement does not clear the round trip", () => {
    // This is the case that matters: a challenger that looks better but not better ENOUGH. The
    // shipped ranking switches here, and pays $20 per $10k for the privilege.
    expect(shouldSwitch(0.003, 0.0034, band)).toBe(false);
  });

  test("switches when it clearly does", () => {
    expect(shouldSwitch(0.003, 0.004, band)).toBe(true);
  });

  test("a tie holds, because holding is the cheaper error", () => {
    expect(shouldSwitch(0.003, 0.003 + band, band)).toBe(false);
  });

  test("takes any real challenger when there is no incumbent", () => {
    expect(shouldSwitch(Number.NaN, 0.001, band)).toBe(true);
    expect(shouldSwitch(Number.NaN, Number.NaN, band)).toBe(false);
  });
});

describe("shouldExit", () => {
  test("leaves a decaying pair even though no challenger cleared the entry band", () => {
    // Without this, hysteresis holds a dying position forever: the incumbent's score collapses, and
    // precisely because everything is bad, nothing beats it by enough to trigger a switch.
    expect(shouldExit(0.0001, PRIOR)).toBe(true);
    expect(shouldExit(0.003, PRIOR)).toBe(false);
  });

  test("it is easier to leave than to switch", () => {
    const band = switchBand(0.002, 4);
    const decayed = 0.0002;
    // No challenger clears the entry band...
    expect(shouldSwitch(decayed, decayed + band / 2, band)).toBe(false);
    // ...but the exit floor drops it anyway.
    expect(shouldExit(decayed, PRIOR)).toBe(true);
  });

  test("an unscoreable incumbent is exited, not held", () => {
    expect(shouldExit(Number.NaN, PRIOR)).toBe(true);
  });
});

describe("deployableUsd", () => {
  test("2% of the thinner leg, which is the pre-registered default", () => {
    expect(deployableUsd(25_000_000)).toBe(500_000);
    expect(deployableUsd(5_000_000)).toBe(100_000);
  });

  test("the measured thresholds for this segment hold", () => {
    // $100k needs a $5M thinner leg at 2% -- 13 pairs cleared that on 2026-09-13.
    expect(deployableUsd(5_000_000, DEFAULT_PARTICIPATION)).toBeGreaterThanOrEqual(100_000);
    // A $4M leg does not.
    expect(deployableUsd(4_000_000, DEFAULT_PARTICIPATION)).toBeLessThan(100_000);
  });

  test("participation is per client", () => {
    expect(deployableUsd(25_000_000, 0.01)).toBe(250_000);
    expect(deployableUsd(25_000_000, 0.05)).toBe(1_250_000);
  });

  test("no depth means nothing can be deployed", () => {
    expect(deployableUsd(0)).toBe(0);
    expect(deployableUsd(Number.NaN)).toBe(0);
  });
});

describe("rankByExpectedDollars", () => {
  test("a deep modest pair outranks a shallow spectacular one", () => {
    // The inversion this segment needs, in one assertion. 300% APR good for $10k is noise to a
    // $1M-$10M book; 14% good for $600k is the business.
    const ranked = rankByExpectedDollars(
      [
        {
          row: "shallow-spectacular",
          evidence: evidence({ ratePerWeek: 0.0577, thinnerLegOiUsd: 500_000 }),
        },
        {
          row: "deep-modest",
          evidence: evidence({ ratePerWeek: 0.0027, thinnerLegOiUsd: 30_000_000 }),
        },
      ],
      PRIOR,
    );

    expect(ranked[0]?.row).toBe("deep-modest");
    expect(ranked[0]?.deployableUsd).toBe(600_000);
    // And the rate ordering is the opposite, which is exactly the point.
    expect(ranked[1]?.score).toBeGreaterThan(ranked[0]?.score as number);
  });

  test("reports dollars and rate side by side, because both are needed", () => {
    const [top] = rankByExpectedDollars([{ row: "x", evidence: evidence({}) }], PRIOR);
    expect(top?.deployableUsd).toBe(500_000);
    expect(top?.expectedWeeklyUsd).toBeCloseTo((top?.score as number) * 500_000, 6);
  });

  test("the order is total, so it does not churn between runs on a tie", () => {
    // An unstable sort would reintroduce turnover through the back door.
    const rows = [
      { row: "a", evidence: evidence({}) },
      { row: "b", evidence: evidence({}) },
      { row: "c", evidence: evidence({}) },
    ];
    const first = rankByExpectedDollars(rows, PRIOR).map((r) => r.expectedWeeklyUsd);
    const again = rankByExpectedDollars(rows, PRIOR).map((r) => r.expectedWeeklyUsd);
    expect(first).toEqual(again);
  });

  test("a pair with no depth ranks last rather than being dropped", () => {
    // Dropping it would hide it; ranking it at zero dollars says why it is useless.
    const ranked = rankByExpectedDollars(
      [
        { row: "no-depth", evidence: evidence({ ratePerWeek: 0.05, thinnerLegOiUsd: 0 }) },
        { row: "real", evidence: evidence({}) },
      ],
      PRIOR,
    );
    expect(ranked[1]?.row).toBe("no-depth");
    expect(ranked[1]?.expectedWeeklyUsd).toBe(0);
  });
});

describe("the churn arithmetic this module exists to fix", () => {
  test("switching 71% of a book weekly costs about half the gross return", () => {
    // The measured case: 0.71 of asset slots re-selected per run, 20 bps per round trip, against a
    // 0.003/week gross. This is the number that justifies the band.
    const drag = 0.71 * 0.002; // per week
    const gross = 0.003; // per week
    expect(drag * 52).toBeCloseTo(0.0738, 4); // 7.4% of capital a year
    expect(gross * 52).toBeCloseTo(0.156, 3); // 15.6% APR
    expect(drag / gross).toBeGreaterThan(0.45);
    expect(drag / gross).toBeLessThan(0.5);
  });

  test("the band would have blocked a switch of the size that churns", () => {
    // A challenger better by a tenth of the weekly rate does not clear payback on a four-week hold.
    expect(shouldSwitch(0.003, 0.003 * 1.1, switchBand(0.002, 4))).toBe(false);
  });
});

File: apps/collector-go/internal/core/ranking_test.go
package core

import (
	"math"
	"reflect"
	"testing"
)

// Vectors from packages/core/src/ranking.test.ts.

// rkEvidence is the measured shape of a well-evidenced pair.
func rkEvidence(over func(*PairEvidence)) PairEvidence {
	e := PairEvidence{RatePerWeek: 0.003, ChargeDays: 30, ThinnerLegOiUSD: 25_000_000}
	if over != nil {
		over(&e)
	}
	return e
}

// rkPrior is the measured median of $3.12 per $10k per week.
const rkPrior = 0.000312

func TestScoreShrinksThinSamplesTowardThePrior(t *testing.T) {
	sparse := Score(rkEvidence(func(e *PairEvidence) { e.RatePerWeek = 0.01; e.ChargeDays = 4 }), rkPrior, ShrinkageK)
	thick := Score(rkEvidence(func(e *PairEvidence) { e.RatePerWeek = 0.01; e.ChargeDays = 30 }), rkPrior, ShrinkageK)
	assertClose(t, "sparse", sparse, (4*0.01+10*rkPrior)/14, 8)
	assertClose(t, "thick", thick, (30*0.01+10*rkPrior)/40, 8)
	if !(sparse < thick) {
		t.Errorf("sparse %v should score below thick %v", sparse, thick)
	}
	if !(sparse < 0.01*0.75) {
		t.Errorf("sparse %v should be pulled most of the way to the prior", sparse)
	}
}

func TestScoreWithNoEvidenceIsThePrior(t *testing.T) {
	assertClose(t, "no evidence", Score(rkEvidence(func(e *PairEvidence) { e.ChargeDays = 0 }), rkPrior, ShrinkageK), rkPrior, 10)
}

func TestScorePriorIsTheMedianNotZero(t *testing.T) {
	got := Score(rkEvidence(func(e *PairEvidence) { e.RatePerWeek = 0; e.ChargeDays = 1 }), rkPrior, ShrinkageK)
	if !(got > 0) {
		t.Errorf("score %v, want > 0", got)
	}
}

func TestScoreNaNInNaNOut(t *testing.T) {
	if got := Score(rkEvidence(func(e *PairEvidence) { e.RatePerWeek = math.NaN() }), rkPrior, ShrinkageK); !math.IsNaN(got) {
		t.Errorf("score %v, want NaN", got)
	}
}

func TestSwitchBandPaysForItselfOverTheHold(t *testing.T) {
	assertClose(t, "band", SwitchBand(0.002, 4, BandCubeRootC), 0.0005, 10)
}

func TestSwitchBandIsAMeaningfulFractionOfTheRate(t *testing.T) {
	ratio := SwitchBand(0.002, 4, BandCubeRootC) / 0.003
	if !(ratio > 0.1 && ratio < 0.3) {
		t.Errorf("band/rate %v, want between 0.1 and 0.3", ratio)
	}
}

func TestSwitchBandBetterFeeTierLowersTheBar(t *testing.T) {
	if !(SwitchBand(0.0006, 4, BandCubeRootC) < SwitchBand(0.002, 4, BandCubeRootC)) {
		t.Error("a cheaper round trip should lower the band")
	}
}

func TestSwitchBandCubeRootTermIsOffUntilCalibrated(t *testing.T) {
	if SwitchBand(0.002, 4, BandCubeRootC) != SwitchBand(0.002, 4, 0) {
		t.Error("the default must equal c = 0")
	}
	if !(SwitchBand(0.002, 4, 0.01) > SwitchBand(0.002, 4, 0)) {
		t.Error("a positive c should widen the band")
	}
}

func TestSwitchBandNoCostNoBand(t *testing.T) {
	if SwitchBand(0, DefaultHorizonWeeks, BandCubeRootC) != 0 || SwitchBand(math.NaN(), DefaultHorizonWeeks, BandCubeRootC) != 0 {
		t.Error("zero or NaN cost must give a zero band")
	}
}

func TestShouldSwitch(t *testing.T) {
	band := SwitchBand(0.002, 4, BandCubeRootC)
	if ShouldSwitch(0.003, 0.0034, band) {
		t.Error("holds when the improvement does not clear the round trip")
	}
	if !ShouldSwitch(0.003, 0.004, band) {
		t.Error("switches when it clearly does")
	}
	if ShouldSwitch(0.003, 0.003+band, band) {
		t.Error("a tie holds")
	}
	if !ShouldSwitch(math.NaN(), 0.001, band) {
		t.Error("takes any real challenger when there is no incumbent")
	}
	if ShouldSwitch(math.NaN(), math.NaN(), band) {
		t.Error("no real challenger, no switch")
	}
}

func TestShouldExit(t *testing.T) {
	if !ShouldExit(0.0001, rkPrior) || ShouldExit(0.003, rkPrior) {
		t.Error("leaves a decaying pair and keeps a healthy one")
	}
	band := SwitchBand(0.002, 4, BandCubeRootC)
	decayed := 0.0002
	if ShouldSwitch(decayed, decayed+band/2, band) {
		t.Error("no challenger clears the entry band")
	}
	if !ShouldExit(decayed, rkPrior) {
		t.Error("but the exit floor drops it anyway")
	}
	if !ShouldExit(math.NaN(), rkPrior) {
		t.Error("an unscoreable incumbent is exited")
	}
}

func TestDeployableUSD(t *testing.T) {
	if DeployableUSD(25_000_000, DefaultParticipation) != 500_000 || DeployableUSD(5_000_000, DefaultParticipation) != 100_000 {
		t.Error("2% of the thinner leg")
	}
	if !(DeployableUSD(5_000_000, DefaultParticipation) >= 100_000) || !(DeployableUSD(4_000_000, DefaultParticipation) < 100_000) {
		t.Error("the measured $100k threshold")
	}
	if DeployableUSD(25_000_000, 0.01) != 250_000 || DeployableUSD(25_000_000, 0.05) != 1_250_000 {
		t.Error("participation is per client")
	}
	if DeployableUSD(0, DefaultParticipation) != 0 || DeployableUSD(math.NaN(), DefaultParticipation) != 0 {
		t.Error("no depth deploys nothing")
	}
}

func TestExpectedWeeklyUSD(t *testing.T) {
	assertClose(t, "weekly", ExpectedWeeklyUSD(0.003, 500_000), 1_500, 6)
	if !(ExpectedWeeklyUSD(0.0577, 10_000) < ExpectedWeeklyUSD(0.0027, 600_000)) {
		t.Error("rate alone does not decide")
	}
	if ExpectedWeeklyUSD(math.NaN(), 500_000) != 0 || ExpectedWeeklyUSD(0.003, math.NaN()) != 0 {
		t.Error("zero rather than NaN when either side is unknown")
	}
}

func TestRankByExpectedDollarsDeepModestOutranksShallowSpectacular(t *testing.T) {
	ranked := RankByExpectedDollars([]RankInput[string]{
		{Row: "shallow-spectacular", Evidence: rkEvidence(func(e *PairEvidence) { e.RatePerWeek = 0.0577; e.ThinnerLegOiUSD = 500_000 })},
		{Row: "deep-modest", Evidence: rkEvidence(func(e *PairEvidence) { e.RatePerWeek = 0.0027; e.ThinnerLegOiUSD = 30_000_000 })},
	}, rkPrior, DefaultParticipation)
	if ranked[0].Row != "deep-modest" || ranked[0].DeployableUSD != 600_000 {
		t.Fatalf("top %+v, want deep-modest at $600k", ranked[0])
	}
	if !(ranked[1].Score > ranked[0].Score) {
		t.Error("the rate ordering is the opposite")
	}
}

func TestRankByExpectedDollarsReportsDollarsAndRate(t *testing.T) {
	ranked := RankByExpectedDollars([]RankInput[string]{{Row: "x", Evidence: rkEvidence(nil)}}, rkPrior, DefaultParticipation)
	if ranked[0].DeployableUSD != 500_000 {
		t.Errorf("deployable %v, want 500000", ranked[0].DeployableUSD)
	}
	assertClose(t, "weekly", ranked[0].ExpectedWeeklyUSD, ranked[0].Score*500_000, 6)
}

func TestRankByExpectedDollarsOrderIsStable(t *testing.T) {
	rows := []RankInput[string]{{Row: "a", Evidence: rkEvidence(nil)}, {Row: "b", Evidence: rkEvidence(nil)}, {Row: "c", Evidence: rkEvidence(nil)}}
	values := func() []float64 {
		out := []float64{}
		for _, r := range RankByExpectedDollars(rows, rkPrior, DefaultParticipation) {
			out = append(out, r.ExpectedWeeklyUSD)
		}
		return out
	}
	if first, again := values(), values(); !reflect.DeepEqual(first, again) {
		t.Errorf("order changed between runs: %v then %v", first, again)
	}
	// Stability is also directly visible: equal rows keep their input order.
	ranked := RankByExpectedDollars(rows, rkPrior, DefaultParticipation)
	if ranked[0].Row != "a" || ranked[1].Row != "b" || ranked[2].Row != "c" {
		t.Errorf("tied rows reordered: %v %v %v", ranked[0].Row, ranked[1].Row, ranked[2].Row)
	}
}

func TestRankByExpectedDollarsNoDepthRanksLast(t *testing.T) {
	ranked := RankByExpectedDollars([]RankInput[string]{
		{Row: "no-depth", Evidence: rkEvidence(func(e *PairEvidence) { e.RatePerWeek = 0.05; e.ThinnerLegOiUSD = 0 })},
		{Row: "real", Evidence: rkEvidence(nil)},
	}, rkPrior, DefaultParticipation)
	if ranked[1].Row != "no-depth" || ranked[1].ExpectedWeeklyUSD != 0 {
		t.Errorf("last %+v, want no-depth at 0", ranked[1])
	}
}

func TestChurnArithmetic(t *testing.T) {
	drag := 0.71 * 0.002
	gross := 0.003
	assertClose(t, "drag", drag*52, 0.0738, 4)
	assertClose(t, "gross", gross*52, 0.156, 3)
	if ratio := drag / gross; !(ratio > 0.45 && ratio < 0.5) {
		t.Errorf("drag/gross %v, want between 0.45 and 0.5", ratio)
	}
	if ShouldSwitch(0.003, 0.003*1.1, SwitchBand(0.002, 4, BandCubeRootC)) {
		t.Error("the band blocks a switch of the size that churns")
	}
}
